### DISCUSSION TITLE
Is it possible to specify files to fix?

### DISCUSSION THREAD

**Discussion by piljac1** [PHPCSStandards/PHP_CodeSniffer#322] [Q&A] [answered]
Hi everyone! I'm relatively new to phpcs since in the past, I only relied on php-cs-fixer, but now, our team wants more detailed linting (even for non automatically fixable issues).

Working with php-cs-fixer, you can specify specific files to fix (alongside an intersection argument to make sure you only fix files that match your config). This is really handy for targeted linting/fixing, such as pre-push (or pre-commit) hooks and pull request pipelines (only lint files that are part of the PR).

Now I tried to look around PHP CodeSniffer docs to try and find the same feature, but to no avail. It seems like the most popular approach is oriented towards file/directory exclusion rather than inclusion/intersection.

Does this "lint only the specified files" feature exist on PHP CodeSniffer?

Thanks in advance!

**Comment by stronk7** [comment, marked as answer]
Hi,

while I've not tried many of them, it seems that there are various alternatives to be able to effectively define the exact files to be checked. The `--help` shows a bunch of them:

https://github.com/phpcsstandards/PHP_CodeSniffer/wiki/Usage

- The usual list of multiple files and folders (`phpcs /path/to/this.php /path/to/that/dir ...`).
- A list of files and directories stored in a file (`phpcs --file-list=/path/to/file_list.txt`).
- A custom filter class where you can implement your "finder/filtering rules" (`phpcs --filter=/path/to/filter/class`).

I imagine that with some of those alternatives, maybe together with `--ignore`, you can accurately define the exact files/dirs to be checked without too much trouble.

Hope it helps, as commented above I haven't used much any of those alternatives, but they look like viable alternatives.

Ciao :-)

PS: What we use to do in our CI jobs (don't know if that's your case) is to proceed to remove all the unmodified files, only keeping the modified files corresponding to the branch being checked (`git diff --stat` basically). While it's sort of "manual solution", it works pretty well and ensures that only what is modified is checked and not the whole codebase that is huge. Not saying that it's the best solution, but for CI purposes (disposable results)... it keeps things simple (it's easy to delete everything but the target files). In fact, we also remove everything not related with the exact lines being modified, but that's another story, heh.

**Reply by piljac1** [reply]
Pretty embarrassing that I missed this page! I checked the Avanced Usage one and didn't see the regular Usage one for some reason... Oh well! Thanks for your answer, I'll try it out in the next couple of days 😄 

And regarding your PS, I use `git diff --name-only --diff-filter=ACMRTUXB` followed by the origin (remote) version of the branch or a target branch and source branch in case of a pull request pipeline 😃

**Reply by jrfnl** [reply]
In addition to what @stronk7 said:

> This is really handy for targeted linting/fixing, such as pre-push (or pre-commit) hooks

There are predefined filters available for this: `--filter=GitModified`/`--filter=GitStaged`.

> What we use to do in our CI jobs is to proceed to remove all the unmodified files, only keeping the modified files corresponding to the branch being checked (git diff --stat basically)

I'd love to add a `BranchModified` filter too, but it needs the name of the branch the current branch branched off from to determine an accurate diff, which is why that gets complicated, so creating that `diff` yourself and then passing the file names to PHPCS is what I know quite some others do too.

There is also extensive [documentation available in the Wiki](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki), including all sorts of tips and tricks on how to [configure your custom ruleset ](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Annotated-Ruleset)exactly how you want it.
In the custom ruleset, you can, for instance, configure certain sniffs to only run for `phpcs` (checking) and not for `phpcbf` (fixing) and visa versa.

Hope you enjoy using PHP_CodeSniffer!

**Reply by stronk7** [reply]
> And regarding your PS, I use `git diff --name-only --diff-filter=ACMRTUXB` followed by the origin (remote) version of the branch or a target branch and source branch in case of a pull request pipeline.

Just in case it can be of useful for anybody, this is the script that we use to  generates an easy to read XML file from any given unified diff (it has some - `clilib.php` dependency, but easy to solve - the class itself works standalone:

https://github.com/moodlehq/moodle-local_ci/blob/main/diff_extract_changes/diff_extract_changes.php

Then, it's easy to use it to decide which files to process and, more yet, which are the warnings/errors (by line) to report later.

Ciao :-)

**Comment by jrfnl** [comment]
FYI: While it doesn't directly address this issue, PR #447 is a proposal to improve the help screens, including improved information on the help screen about how to specify which files should be scanned.

Reviews and testing of that PR would be appreciated.
